Add HSM setup docs (#31)
alphagov/gsp-team-manual: docs/setting-up-hsm.md

EOF
```

The `openssl` command mentions several files:
- `customerCA.crt` is basically the CA certificate exported from the key.
  Easily obtainable from GUI or the `yubico-piv-tool`
- `cluster-rn3c7izgvya_ClusterCsr.csr` is the file we obtain from HSM itself,
  following [these instructions](https://docs.aws.amazon.com/cloudhsm/latest/userguide/initialize-cluster.html#get-csr)
- `cluster-rn3c7izgvya_CustomerHsmCertificate.crt` will be generated upon
  successful signing

The file names are bit random. They represent the HSM's ID.

In general, this post was better than any other documentation I found:
https://dennis.silvrback.com/openssl-ca-with-yubikey-neo

After we have signed the certificate, we can continue with the initialisation
and activation of the cluster. We should be following the AWS guide on any of
this, except the certificate signing which we do our way.

### Push the customerCA.crt to AWS SSM

This signed customer certificate will need to be shared with clients
interacting with the HSM.

We decided to store it along the passwords in HSM for easier access.

```sh

> aws ssm put-parameter --type SecureString --name /hsm/customerCA --value "$(cat customerCA.crt)"

Reviewer: We need to be wary that this may not always be possible. `SecureString`s have limits. According to the [docs](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#API_PutParameter_RequestSyntax) the limit is 4096, which seems plenty big enough. But we defnitely hit that limit with SSH keys in the past.

# Setting up HSM

We're intending to use hardware security modules (HSM) as the means to sign
given piece of information with non-extractable key.

At the moment, we're committing to use CloudHSM provided by AWS.

> As it turns out, it's not a straight forward task to set this up.

Author: ```suggestion
```

x509 -engine pkcs11 -CAkeyform engine -CAkey slot_0-id_2 -CAcreateserial -sha256 -CA hsm.gds.crt -req -in cluster-rn3c7izgvya_ClusterCsr.csr -out cluster-rn3c7izgvya_CustomerHsmCertificate.crt
EOF
```

The `openssl` command mentions few files:
- `hsm.gds.crt` is basically the public certificate exported from the key.
  Easily obtainable from GUI or the `yubico-piv-tool`
- `cluster-rn3c7izgvya_ClusterCsr.csr` is the file we obtain from HSM itself,
  following [these instructions](https://docs.aws.amazon.com/cloudhsm/latest/userguide/initialize-cluster.html#get-csr)
- `cluster-rn3c7izgvya_CustomerHsmCertificate.crt` will be generated upon
  successful signing

The file names are bit random. They represent the HSM's ID.

In general, this post was better than any other documentation I found:
https://dennis.silvrback.com/openssl-ca-with-yubikey-neo

After we have signed the certificate, we can continue with the initialisation
and activation of the cluster. We should be following the AWS guide on any of
this, except the certificate signing which we do our way.

> ### Push the customerCA.crt to AWS SSM

Author: 👍

This signed customer certificate will need to be shared with clients
interacting with the HSM.

We decided to store it along the passwords in HSM for easier access.

```sh
aws ssm put-parameter --type SecureString --name /hsm/customerCA --value "$(cat hsm.gds.crt)"
```

Obtainable with:

```sh
aws ssm get-parameter --query Parameter.Value --output text --with-decryption --name /hsm/customerCA
```

## Activating the HSM

The complication may be, that the HSM cluster has ingress, meaning we can only
get onto it from specific boxes. K8s EC2 Nodes in our case.

```sh

> kubectl run ubuntu --image ubuntu:trusty sleep 1800

Reviewer: ```suggestion
kubectl run ubuntu -it --rm --image ubuntu:trusty -- bash
```

> 

Reviewer: ```suggestion
```

> kubectl get pods

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> kubectl exec -it ubuntu-7f457cd468-tsmc5 -- bash

Reviewer: ```suggestion
```

## On the box follow AWS guide to install client: https://docs.aws.amazon.com/cloudhsm/latest/userguide/install-and-configure-client-linux.html

## Follow the Activate Cluster AWS docs: https://docs.aws.amazon.com/cloudhsm/latest/userguide/activate-cluster.html
```

The activation of the cluster involves password changing and user creation.
Once again, [AWS docs](https://docs.aws.amazon.com/cloudhsm/latest/userguide/activate-cluster.html)
and client `help` should be sufficient to accomplish that.

The passwords are generated on Programme's AWS SSM, with the use of:

```sh

> aws ssm put-parameter --type SecureString --name /hsm/users/co/password --value "$(pwgen -c -n -s -y -1 -B 32 1 | tr -d '\n')" ## For Crypto Officer

Reviewer: ```suggestion
aws ssm put-parameter --type SecureString --name /hsm/users/co/password --value "$(pwgen --capitalize --numerals --secure --symbols -1 --ambiguous 32 1 | tr -d '\n')" ## For Crypto Officer
```

> aws ssm put-parameter --type SecureString --name /hsm/users/cu/1/password --value "$(pwgen -c -n -s -y -1 -B 32 1 | tr -d '\n')" ## For Crypto User - Increased by 1 for any next user.

Reviewer: ```suggestion
aws ssm put-parameter --type SecureString --name /hsm/users/cu/1/password --value "$(pwgen --capitalize --numerals --secure --symbols -1 --ambiguous 32 1 | tr -d '\n')" ## For Crypto User - Increased by 1 for any next user.
```

```

> These passwords then can be obtainable by:

Reviewer: ```suggestion
These passwords can be retrieved with:
```
